pkg/ingester/client/streaming.go

```diff
@@ -141,15 +144,20 @@ func (s *SeriesChunksStreamReader) readStream(log *spanlogger.SpanLogger) error
 		}
 
 		select {
-		case <-s.client.Context().Done():
+		case <-s.ctx.Done():
 			// Why do we abort if the context is done?
 			// We want to make sure that this goroutine is never leaked.
 			// This goroutine could be leaked if nothing is reading from the buffer, but this method is still trying to send
 			// more series to a full buffer: it would block forever.
 			// So, here, we try to send the series to the buffer if we can, but if the context is cancelled, then we give up.
 			// This only works correctly if the context is cancelled when the query request is complete or cancelled,
 			// which is true at the time of writing.
-			return s.client.Context().Err()
+			//
+			// Note that we deliberately don't use the context from the gRPC client here: that context is cancelled when
+			// the stream's underlying ClientConn is closed, which can happen if the querier decides that the ingester is no
+			// longer healthy. If that happens, we want to return the more informative error we'll get from Recv() above, not
+			// a generic 'context canceled' error.
+			return s.ctx.Err()
```

Comment:
can we wrap the error with something? It will make context cancelled errors easier to trace

also, if we use `context.Cause(ctx)` we might get some more details in the future

same comment applies to the store-gateway changes

Reply:
> can we wrap the error with something? It will make context cancelled errors easier to trace

Good idea, I've added this in 7fdff2d. The errors would have been wrapped in a message like `attempted to read series at index X from ingester chunks stream, but the stream has failed` in `GetChunks()` below, but adding more details will help isolate the source to this line, so makes sense to me.

> also, if we use `context.Cause(ctx)` we might get some more details in the future
TIL, thanks! I wonder if we should set up a linting rule to prefer the use of `context.Cause(ctx)` over `ctx.Err()`.